### ISSUE TITLE
Documentation around GIT_DESCRIBE environment variables 

### ISSUE THREAD

**Issue by mrocklin** [conda/conda-build#322]
Hi all, 

I've tried naively following the instructions surrounding the use of `git describe` environment variables found here http://conda.pydata.org/docs/build.html . My solution is here https://github.com/mrocklin/into/blob/conda-build-recipe/conda.recipe/meta.yaml

However when I run the following:

    ~/workspace/into/$ conda build conda.recipe

I get the following error:

```
  File "/home/mrocklin/Software/anaconda/lib/python2.7/site-packages/conda_build/metadata.py", line 311, in build_number
    return int(self.get_value('build/number', 0))
TypeError: int() argument must be a string or a number, not 'NoneType'
```

In a small attempt to locate the issue I inspect what's happening in `get_git_build_info`. I find oddly that the `git_dir` and `src_dir` doesn't point to my repository, but rather to something in `conda-bld`

```
> /home/mrocklin/Software/anaconda/lib/python2.7/site-packages/conda_build-1.10.0_117_g807392d_dirty-py2.7.egg/conda_build/environ.py(37)get_git_build_info()
-> if os.path.isdir(git_dir):
(Pdb) print git_dir
/home/mrocklin/Software/anaconda/conda-bld/work/.git
```

Probably I'm doing something dumb.  But it's unclear to me what that dumb thing is.

**Comment by melund** [comment]
I get the same error. Maybe something is broken in the newest version of conda-build... 

Downgrading to conda-build 1.10.0 solves the problem for me.

**Comment by asmeurer** [comment]
There are two issues at play here:

The GIT_* environment variables are set to empty strings when the source is not a git repository. This behavior is wrong given the `environ.get()` code that we recommend putting in people's meta.yaml. They should not be set at all, so that the second argument of the get is used, rather than being set to None (an empty value in a dictionary in yaml is implicitly null). This is easy to fix, and I will do so now. The fix for this should make your recipe work. 

The second issue is that the source directory in the `get_git_build_info` function is not a git repository yet when it is run, because it hasn't been cloned yet. @tswicegood I know we had issues like this before. Do you remember what the status of this is?

**Comment by tacaswell** [comment]
There is a somewhat more pathological case where the source directory contains a git checkout from the _last_ build which may be an entirely different project.  In the cases where the previous build has tag names that are not valid version strings (which can be handled in the template if you know its coming) which will cause your build to fall over and die with a tremendously unhelpful error message ('not-thepackge-you-are-bulidng version is not a valid version string').  In those cases the solution is to find and manually delete the work directory.

Either the build process needs to clean up _before_ the first templating is run or the templating should be skipped on the first read of the yaml (and only extract the source field).

**Comment by asmeurer** [comment]
We obviously can't clone the git repo before the meta.yaml is read, but I think we probably can clear the work directory.  

I'm not really sure what skipping the templating would mean.

**Comment by asmeurer** [comment]
Something like [this](https://github.com/conda/conda-build/pull/325). It's not really a fix, but it should avoid behavior that's difficult to reproduce because it depends on the previous build.

**Comment by tacaswell** [comment]
Just read the `meta.yaml` file with out running it through jinja + validation.

**Comment by asmeurer** [comment]
Unfortunately, that's not possible. The jinja2 templating markers are not valid YAML.

**Comment by tswicegood** [comment]
@tacaswell skipping jinja parsing would also mean that some packages wouldn't have a name.  For example, the [`load_setuptools`][1] code can be used to pull information like version and so on from the `setup.py` file.

I do agree we need to have a more clear way to handle the parsing, I just haven't figured out a good way to do it.

[1]: https://github.com/conda/conda-build/blob/master/conda_build/jinja_context.py#L17-L36

**Comment by asmeurer** [comment]
I think the only way to really fix this would be to create special metavariables for common things, like GIT_DESCRIBE_TAG, and use those directly in the meta.yaml. Then conda-build could replace those with the real values at the appropriate times and not replace them at inappropriate times.  

@tacaswell's suggestion is actually not too bad, except for the issue that the jinja2 template markers are invalid yaml. The only way around that is to require users to quote the templating all the time, like `string: "{{ environ.get('GIT_BUILD_STR', '') }}"` (but even then there are annoying issues to deal with if you need to escape the quotes).

**Comment by tacaswell** [comment]
Can you do a sed-like brute force clean up frist? Something like `s/{{.*}}/' '/g` to strip all the jinja mark up.

**Comment by tacaswell** [comment]
To be more concrete, something like:

```python

import re
import yaml

def safe_yaml_read(fpath, replace_str=''):
    """
    Reads a yaml file stripping all of the jinja templating markup

    Parameters
    ----------
    fpath : str
        Path to yaml file to sanitize

    replace_str : str
        String to replace the template markup with, defaults to ''.

    Returns
    -------
    yaml_dict : dict
        The dictionary with all of the jinja2 templating fields
        replaced with ``replace_str``.
    """
    with open(fpath, 'r') as f:
        lns = []
        for ln in f:
            lns.append(re.sub(r'{[{%].*?[%}]}', '', ln))
    meta_dict = yaml.load(''.join(lns))
    return meta_dict

```

**Comment by stuarteberg** [comment]
Unless I'm mistaken, this issue will be resolved when #666 is merged.  From the discussion above:

>Just read the meta.yaml file with out running it through jinja + validation.

...
>Unfortunately, that's not possible. The jinja2 templating markers are not valid YAML.

The solution in #666 is to tell jinja to ignore undefined values during the first pass (just replace them with empty strings).  It's similar to the "sed-like" suggestion mentioned above, except we just let jinja do it for us, and only for the templates that have undefined variables.

**Comment by msarahan** [comment]
I think this is resolved in the latest release (1.20.2beta).  If there is still confusion, please reopen this issue, or open a new one.
